File: fixtures/project/spec-verify-jsonrpc/python/jsonrpc/merge_patch.py
"""JSON Merge Patch (RFC 7396) implementation."""

from typing import Any


def merge_patch(target: Any, patch: Any) -> Any:
    """Apply a JSON Merge Patch to a target document."""
    if isinstance(patch, dict):
        if not isinstance(target, dict):
            target = {}
        for name, value in patch.items():
            if value is None:
                target.pop(name, None)
            else:
                target[name] = merge_patch(target.get(name), value)
        return target
    return patch

File: fixtures/project/spec-verify-jsonrpc/typescript/src/merge-patch.ts
/**
 * Apply a JSON Merge Patch (RFC 7396) to a target document.
 */
export function mergePatch(target: unknown, patch: unknown): unknown {
  if (
    patch !== null &&
    typeof patch === "object" &&
    !Array.isArray(patch)
  ) {
    let result: Record<string, unknown>;
    if (
      target !== null &&
      typeof target === "object" &&
      !Array.isArray(target)
    ) {
      result = { ...(target as Record<string, unknown>) };
    } else {
      result = {};
    }
    for (const [name, value] of Object.entries(
      patch as Record<string, unknown>
    )) {
      if (value === null) {
        delete result[name];
      } else {
        result[name] = mergePatch(result[name], value);
      }
    }
    return result;
  }
  return patch;
}
